[REEF-1224] IMRU Fault Tolerance - Separate Data downloading from Tas…

dkm2110 opened the pull request:
…k injection

This addressed the issue by
  * forcing IDataPartition to be instantiated at root context and service level in mappers
  * instantiating the Network service at task level to make fault tolerance easier

JIRA:
  [REEF-1224](https://issues.apache.org/jira/browse/REEF-1224)

dkm2110 commented:
@markusweimer @jwang98052  this is the new pull request that should replace https://github.com/apache/reef/pull/939. Sorry for the inconvenience.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/DataLoadingContext.cs:
```diff
@@ -0,0 +1,50 @@
+
+        public void OnError(Exception error)
+        {
+            Exceptions.Throw(error, "Error occured in Data Loading context start", Logger);
```
I don't think this ever gets called.

dkm2110 replied:
will remove

dkm2110 replied:
This is a mandatory function for IObserver<> . I cannot remove it.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/DataLoadingContext.cs:
```diff
@@ -0,0 +1,50 @@
+        private static readonly Logger Logger = Logger.GetLogger(typeof(DataLoadingContext<T>));
+
+        [Inject]
+        internal DataLoadingContext(IInputPartition<T> partition)
```
`private`.

dkm2110 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/DataLoadingContext.cs:
```diff
@@ -0,0 +1,50 @@
+        }
+
+        public void OnNext(IContextStart value)
+        {
```
Sorry, I haven't really been following the entire conversation, but should this function do something?

dkm2110 replied:
Eventually yes, once REEF-1339 gets addressed this will call AddCache() funtion introduced in IInputPartition()

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -53,32 +54,40 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
-        private bool _reachedUpdateTaskActiveContext = false;
         private readonly bool _invokeGC;
+        private bool _imruDone = false;
+        private int _taskReady = 0;
```
what is this variable used for?

dkm2110 replied:
This is more like a logging counter that is used for generating messages that how man tasks are ready.

markusweimer replied:
Maybe use a more expressive name like `_numberOfReadyTasks`?

dkm2110 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,54 +195,111 @@ public void OnNext(IActiveContext activeContext)
-                task.ActiveContext.Dispose();
+                Logger.Log(Level.Info,
+                    string.Format("Received completed task message from task Id: {0}", completedTask.Id));
+                _completedTasks.Add(completedTask);
```
Do we ever re-use the `ActiveContext` of a `CompletedTask`?

dkm2110 replied:
No. You are probably wondering why am I collecting them right? So at the end of IMRU master sends done message to slaves/mappers ad then exits the task. I do not want the streams to get closed unless slaves actually receive the message and also reach completed task stage.

afchung replied:
Sounds like you want to be the closing of all Tasks to be synchronized only after the job is done or failed?

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
         /// </summary>
         /// <param name="evaluatorId"></param>
-        internal void EvaluatorFailed(string evaluatorId)
+        /// <returns>Whether failed evaluator is master or not</returns>
```
I'm a bit skeptical about this. Why not leave it as two separate functions?

dkm2110 replied:
u mean have void EvaluatorFailed(string Id) and check for master separately?

afchung replied:
I think this class tries to do way too many things, would be easier to understand if it was broken up -- though I'm not sure if it's a task for this PR.

jwang98052 replied:
This code will be refactored for fault tolerant. I wouldn't spend time of it now.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,54 +195,111 @@ public void OnNext(IActiveContext activeContext)
             }
             else
             {
+                Logger.Log(Level.Info, "Requesting for the failed master evaluator again");
```
The conditions seem to be inverted.

dkm2110 replied:
yupp ......good catch....

markusweimer replied:
Reword `Requesting new master Evaluator`

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,54 +195,111 @@ public void OnNext(IActiveContext activeContext)
-            _serviceAndContextConfigurationProvider.EvaluatorFailed(value.Id);
+            bool isMaster = _serviceAndContextConfigurationProvider.EvaluatorFailed(value.Id);
 
             // if active context stage is reached for Update Task then assume that failed
```
Update the comment?

dkm2110 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -36,31 +34,30 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
     /// <typeparam name="TMapOutput"></typeparam>
-    internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutput>
+    /// <typeparam name="TDataHandler"></typeparam>
+    internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutput, TDataHandler>
```
It seems like this class is doing way more than its name promises -- I think it will be a lot more clear if either its functions were broke down into more classes, or if it was renamed into something that gives more of an insight into what it does.

afchung replied:
Some methods in this class are thread safe, some are not -- what is the thread-safety property of this class?

dkm2110 replied:
Any suggestion for name? I am a little hesitant in doing major refactoring here since this is sort of transient code and will change with fault tolerance. The main aim was to move data loading to context and add group comm. to task.

afchung replied:
Right, I don't know if it's a Task for this PR, just something of note. I'm having a bit of a hard time understanding the code.

dkm2110 replied:
Andrew - Except private functions and constructor everything is thread safe. Can you point me the issue?

afchung replied:
@dkm2110 Ah you're right. Disregard the comment then :) Thanks!

markusweimer replied:
How do we document thread safety?

dkm2110 replied:
So in this case all the operations of this class are thread safe. We at any time are in only one function of this class except for checking for MasterId.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
+        {
+            lock (_lock)
+            {
+                return activeContextId.Equals(IMRUConstants.MasterContextId);
```
Why are we locking on this?

dkm2110 replied:
I think I was copy pasting....no need to lock....will remove

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -36,31 +34,30 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
-            _configurationManager = configurationManager;
-            _assignedPartitionDescriptors = 0;
-            _partitionDescriptors = partitionDescriptors;
+            _contextLoadedEvaluators = new HashSet<string>();
```
Assign at declaration.

dkm2110 replied:
ok

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -36,31 +34,30 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
+            {
+                _partitionDescriptorIds.Push(descriptor.Id);
+            }
             _lock = new object();
```
same.

dkm2110 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -36,31 +34,30 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
-        private readonly Dictionary<string, ContextAndServiceConfiguration> _configurationProvider;
-        private readonly ISet<string> _failedEvaluators;
-        private readonly ISet<string> _submittedEvaluators; 
+        private readonly Dictionary<string, string> _configurationProvider;
```
Why is this named `_configurationProvider`? It doesn't seem like it provides any configurations.

dkm2110 replied:
ok will change to _partitionIdProvider or something

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -36,31 +34,30 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
-            ConfigurationManager configurationManager, Stack<IPartitionDescriptor> partitionDescriptors)
+        private readonly Stack<string> _partitionDescriptorIds;
+        private readonly IPartitionedInputDataSet _dataset;
+        private string _masterEvaluatorId;
```
I would prefer `Optional` in cases where such variables can be `null` and/or reassigned.

dkm2110 replied:
Can you elaborate....is Optional some sort of Keyword?

dkm2110 replied:
The logic here is that if id is null this means either its the beginning and master evaluator isnot registered with this class or the master has failed and we are waiting for another one to come up.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
         /// <param name="evaluatorId"></param>
-        /// <returns></returns>
-        internal ContextAndServiceConfiguration GetNextConfiguration(string evaluatorId)
+        internal void ReachedActiveContext(string evaluatorId)
```
This seems like what a state machine would do... would it be better to modify the collection of Evaluators as `evaluatorId -> EvaluatorState`?

dkm2110 replied:
Yes then it starts looking much more like what is being proposed in Fault tolerant design and will be done in near future...So I am also a bit confused here.

markusweimer replied:
Agreed, let's not get into the state machine business just yet in this PR.

However, the method name needs to change, e.g. `RecordActiveContextPerEvaluatorId()`?

dkm2110 replied:
Changed

afchung commented:
I'm done with a first pass. @markusweimer feel free to have a pass.

dkm2110 commented:
@afchung  I agree with your major higher level issues/comments. However, I was not sure whether to do all that in this PR or not given @jwang98052  wanted to do some of these refactorings/changes at her end. @markusweimer please share your thoughts also at some of @afchung  concerns.

afchung commented:
Thanks @dkm2110, I think you and @markusweimer are better judges of that, given that I'm not too familiar with the code yet. I'll do a pass on the logic after @markusweimer's review and your next revision. Thanks!

markusweimer commented:
Doing a pass now.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/API/IIMRUClient.cs:
```diff
@@ -22,16 +22,17 @@
         /// <param name="jobDefinition">IMRU job definition</param>
         /// <returns>Result of IMRU</returns>
-        IEnumerable<TResult> Submit<TMapInput, TMapOutput, TResult>(IMRUJobDefinition jobDefinition);
+        IEnumerable<TResult> Submit<TMapInput, TMapOutput, TResult, TDataHandler>(IMRUJobDefinition jobDefinition);
```
What is `TDataHandler`?

dkm2110 replied:
The data handler type returned by GetPartitionHandler() in IInputPartition

afchung replied:
It doesn't seem like the type information is used at all, do we actually need it in the `Submit` function?

dkm2110 replied:
This information is used to instanitate IInputPartition in DataLoadingContext.

jwang98052 replied:
Is it something like IEnumerable<Row>? You may want to use another name. It represents data, not a handler.

dkm2110 replied:
It is Data Handler....Since in IInputPartition the corresponding function is GetPartitionHandle. So I would stick with it.

markusweimer replied:
How about `TPartitionType`? `Handler` and `Manager` are examples of heavily overloaded terms, which make it hard to know what they really mean when reading an API.

markusweimer replied:
Also, do we really need to introduce these types all the way to the `Submit` method? I struggled with this myself when making the IMRU API. @afchung do you see a way to avoid this pollution of the basic APIs?

dkm2110 replied:
done the name changing....I also do not like this but do not know the way around.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/DataLoadingContext.cs:
```diff
@@ -0,0 +1,50 @@
+        [Inject]
+        internal DataLoadingContext(IInputPartition<T> partition)
+        {
+            Logger.Log(Level.Info, "Entered data loading context");
```
Put this into `Verbose`

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/DataLoadingContext.cs:
```diff
@@ -0,0 +1,50 @@
+            Logger.Log(Level.Info, "Entered data loading context");
+        }
+
+        public void OnNext(IContextStart value)
```
Add documentation to the method as well as a `//TODO[REEF-1339]` comment.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUConstants.cs:
```diff
@@ -25,5 +25,6 @@ internal static class IMRUConstants
         internal const int TreeFanout = 2;
         internal const string MapTaskPrefix = "IMRUMap";
         internal const string UpdateTaskName = "IMRUMaster";
+        internal const string MasterContextId = "MasterRootContext";
```
Why not just `RootContext`?

dkm2110 replied:
This is use only for master.....thats why....Mapper context Ids are based on partition Ids.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -103,23 +111,24 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult> : IObserver<IDr
             _perMapperConfigs = perMapperConfigs;
-            _completedTasks = new ConcurrentBag<ICompletedTask>();
+            _totalMappers = dataSet.Count;
+            _completedTasks = new List<ICompletedTask>();
```
these initializers could move to the definition of the attribute.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -103,23 +111,24 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult> : IObserver<IDr
+            _totalMappers = dataSet.Count;
+            _completedTasks = new List<ICompletedTask>();
+
             _allowedFailedEvaluators = (int)(failedEvaluatorsFraction * dataSet.Count);
```
This rounds down. So if `failedEvaluatorsFraction * dataSet.Count` is less than `1`, it means that no Evaluator is allowed to fail. Is this documented where the `failedEvaluatorsFraction` is defined?

dkm2110 replied:
Will verify and update if not there

dkm2110 replied:
I checked...the definition above is mentioned.,

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,102 +145,46 @@ public void OnNext(IDriverStarted value)
-                        "User has not given any way to handle IMRU result, defaulting to ignoring it");
-                }
-
+                Logger.Log(Level.Info, "Submitting master task");
```
`Verbose`.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,102 +145,46 @@ public void OnNext(IDriverStarted value)
-                        .BindNamedParameter(typeof(InvokeGC), _invokeGC.ToString())
-                        .Build();
-
+                Logger.Log(Level.Info, "Submitting map task");
```
`Verbose`

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,102 +145,46 @@ public void OnNext(IDriverStarted value)
+                    GetMapTaskConfiguration(activeContext, taskId));
+                _groupCommTaskStarter.QueueTask(clientTaskConfig, activeContext);
+                Interlocked.Increment(ref _taskReady);
+                Logger.Log(Level.Info, string.Format("{0} Tasks are ready for submission", _taskReady));
```
`Verbose`

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,54 +195,111 @@ public void OnNext(IActiveContext activeContext)
+
+                if (_completedTasks.Count != _totalMappers + 1)
+                {
+                    return;
```
why not put the remainder of the method into an if block of `if (_completedTasks.Count >= _totalMappers + 1)`?

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,54 +195,111 @@ public void OnNext(IActiveContext activeContext)
+                }
+
+                _imruDone = true;
+                foreach (var task in _completedTasks)
```
Maybe use LINQ?

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,54 +195,111 @@ public void OnNext(IActiveContext activeContext)
-            int currFailedEvaluators = Interlocked.Increment(ref _currentFailedEvaluators);
-
-            if (value.FailedContexts != null && value.FailedContexts.Count != 0)
+            if (_imruDone)
```
You might need a lock around `_imruDone`. Also, maybe find a better name for it?

dkm2110 replied:
you do not need a lock...this is the only place it is set or written and only one Completed task can enter this part of code due to lock in the function.

dkm2110 replied:
I am changing to _imruTasksCompleted

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,54 +195,111 @@ public void OnNext(IActiveContext activeContext)
-            Logger.Log(Level.Info, "Requesting for the failed evaluator again");
-
-            _serviceAndContextConfigurationProvider.EvaluatorFailed(value.Id);
+            bool isMaster = _serviceAndContextConfigurationProvider.EvaluatorFailed(value.Id);
```
This seems a bit too magic. Why not have one call to inform `serviceAndContextConfigurationProvider` of the failed evaluator and another to see whether it was the master?

dkm2110 replied:
Will change to separate calls?

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -307,6 +317,78 @@ public void OnCompleted()
+            if (!_perMapperConfiguration.TryPop(out mapSpecificConfig))
+            {
+                Exceptions.Throw(
+                    new Exception(string.Format("No per map configuration exist for the active context {0}",
```
We need to have better exception types here. This basically indicates a bug, right?

dkm2110 replied:
You mean create new exception classes?

dkm2110 replied:
This indicates a bug in IMRU. Do you want to create something like IMRUBugException sort of class?

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -307,6 +317,78 @@ public void OnCompleted()
         {
         }
 
+        private IConfiguration GetMapTaskConfiguration(IActiveContext activeContext, string taskId)
```
It would help to have some xml docs for the `private` methods added.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
         /// <param name="evaluatorId"></param>
-        internal void EvaluatorFailed(string evaluatorId)
+        /// <returns>Whether failed evaluator is master or not</returns>
+        internal bool EvaluatorFailed(string evaluatorId)
```
This needs a better name. In general, I prefer methods with a verb starting the name. In this case, I'd suggest to: `RecordFailedEvaluatorById()` and `IsMasterEvaluatorId()`.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
+        /// </summary>
+        /// <param name="evaluatorId">Evaluator Id</param>
+        /// <returns>The context and service configuration</returns>
+        internal ContextAndServiceConfiguration GetNextContextConfiguration(string evaluatorId)
```
`GetContextConfigurationForEvaluatorById()`?

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
+        /// </summary>
+        /// <param name="activeContextId">context id</param>
+        /// <returns>true of it is master context id, false otherwise</returns>
+        internal bool IsMasterContext(string activeContextId)
```
In other places, you look these up by the evaluator ID. I am confused as to how we decide which is the master? By Evaluator or Context? Also, shouldn't the group communications topology know this?

If the master is the evaluator / context that runs the update, maybe changing the name to reflect that would help, e.g. `IsIdOfUpdateContext()`?

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
+        /// </summary>
+        /// <param name="evaluatorId"></param>
+        /// <returns></returns>
+        internal string GetPartitionId(string evaluatorId)
```
`GetPartitionIdByEvaluatorId`?

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
+        /// </summary>
+        /// <param name="evaluatorId"></param>
+        /// <returns></returns>
+        private ContextAndServiceConfiguration GetNextDataLoadingConfiguration(string evaluatorId)
```
`GetDataLoadingConfigurationForEvaluator` or `GetDataLoadingConfigurationByEvaluatorId`?

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -69,113 +66,245 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
+            return new ContextAndServiceConfiguration(contextConf, serviceConf);
+        }
+
+        private bool IsMaster(string evaluatorId)
```
There is another method like this. The whole master thing is confusing to me.

dkm2110 replied:
So I can collate to one method probably for checking if it's master. But in general we need to keep tab of master separately from other evaluators right?

markusweimer commented:
Done with my pass.

dkm2110 commented:
@markusweimer  @afchung  I have addressed most of your comments. Please have a look.

afchung commented on lang/cs/Org.Apache.REEF.Bridge/Org.Apache.REEF.Bridge.vcxproj.user:
```diff
@@ -0,0 +1,4 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Is this file relevant?

dkm2110 replied:
no..this added by default if u open VS. Will remove.

afchung replied:
Should we add this to `gitignore`?

afchung commented on lang/cs/Org.Apache.REEF.IMRU.Examples/MapperCount/IdentityMapFunction.cs:
```diff
@@ -26,7 +28,7 @@ namespace Org.Apache.REEF.IMRU.Examples.MapperCount
     {
         [Inject]
-        private IdentityMapFunction()
+        private IdentityMapFunction(IInputPartition<Stream> dataPartition)
```
What is this parameter for?

dkm2110 replied:
I just added it to make sure there is a test and example that takes as input IInputPartition instantiated by DataLoadingContext.

afchung replied:
@dkm2110 How does the test/example verify it?

dkm2110 replied:
@afchung  So partition configuration is now passed only in context and not task. So it ensures that map function has access to partition. But you are right that it does not check whether partition is created twice which would be bug in REEF then.  Do you want me to check that? For this I have to then create another Dataset with static counter in partition constructor which throws error if count exceeds one. Let me know if you want me to introduce it.

jwang98052 replied:
In the current implementation FileSystemInputPartition of IInputPartition, when GetPartitionHandle() is called, we guarantee it is only initialize once. When we add Cache() API, we should make sure data is donwnloaded once with the same IInputPartition instance. 
Our injector for Context and Task must make sure there is only one instance of IInputPartition. Inside task, you should also check it.

jwang98052 replied:
You said you added it to make sure there is a test and example that takes as input IInputPartition instantiated by DataLoadingContext, is it for test purpose? If yes, it should be in test project.

dkm2110 replied:
The reef functional test will throw error otherwise.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/DataLoadingContext.cs:
```diff
@@ -0,0 +1,65 @@
+        /// the caught exception in this case.
+        /// </summary>
+        /// <param name="error">Exception</param>
+        public void OnError(Exception error)
```
These functions never get called, so you don't really need to document/put anything here. They just come with the `IObserver` interface.

dkm2110 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUBugException.cs:
```diff
@@ -0,0 +1,60 @@
+        public IMRUBugException(string message, Exception inner)
+            : base(message, inner)
+        {
+            Logger.Log(Level.Info, "Possible bug in the IMRU code");
```
Don't log here.

dkm2110 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUBugException.cs:
```diff
@@ -0,0 +1,60 @@
+        public IMRUBugException(string message)
+            : base(message)
+        {
+            Logger.Log(Level.Info, "Possible bug in the IMRU code");
```
Don't log here.

dkm2110 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUBugException.cs:
```diff
@@ -0,0 +1,60 @@
+        /// </summary>
+        public override string Message
+        {
+            get { return "Possible Bug in the IMRU code: " + base.Message; }
```
This seems a bit unintuitive. Perhaps you can define a `private static string` function that returns the `const string` appended to `message` and pass it directly to the `base` constructor.

dkm2110 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -53,32 +54,40 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
-    internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult> : IObserver<IDriverStarted>,
-        IObserver<IAllocatedEvaluator>, IObserver<IActiveContext>, IObserver<ICompletedTask>, IObserver<IFailedEvaluator>
+    /// <typeparam name="TDataHandler">Data Handler type in IInputPartition</typeparam>
+    internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TDataHandler> : IObserver<IDriverStarted>,
```
nit: new line for `IObserver<IDriverStarted>`

dkm2110 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -307,6 +315,98 @@ public void OnCompleted()
+                        _configurationManager.UpdateFunctionCodecsConfiguration)
+                    .Build();
+
+            var serviceConf = Configurations.Merge(_groupCommDriver.GetServiceConfiguration(), codecConfig);
```
return directly.

dkm2110 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -307,6 +315,98 @@ public void OnCompleted()
+            {
+                TangFactory.GetTang()
+                    .NewInjector(partialTaskConf, _configurationManager.UpdateFunctionCodecsConfiguration)
+                    .GetInstance<IIMRUResultHandler<TResult>>();
```
Why not just create a default implementation for this Interface?

dkm2110 replied:
I think there is some issue for specifying default implementation for interfaces with generic types. I remember trying but did not work.

afchung replied:
@jwang98052 do you know if this works?

jwang98052 replied:
No, it doesn't work.

jwang98052 replied:
You might want to add some  comments here otherwise every time, people look at this code and will ask this question.

dkm2110 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,54 +194,110 @@ public void OnNext(IActiveContext activeContext)
+                    string.Format("Task with Id: {0} failed but IMRU task is completed. So ignoring.", value.Id));
+                return;
+            }
+            Exceptions.Throw(new Exception(string.Format("Task with Id: {0} failed", value.Id)), Logger);
```
This seems to be inconsistent with your function description.

dkm2110 replied:
fixed the function description

jwang98052 replied:
Do you have test that reaches this line?

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -36,146 +34,264 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
+                IPartitionDescriptor partitionDescriptor =
+                    _dataset.GetPartitionDescriptorForId(_partitionIdProvider[evaluatorId]);
+
+                if (partitionDescriptor == null)
```
This should be the job of the callee, not the caller.

dkm2110 replied:
Removed the exception

afchung commented:
@dkm2110 done with another pass. Overall the changes seem fine for this JIRA item, but I think the documentation and code structure can be improved for IMRU.

markusweimer commented:
One thing @afchung and I wondered off list: This should only need minimal changes on the Driver side, right? Basically, it needs to change the Context and Task Configurations. Why is there this large number of changes to the Driver?

dkm2110 commented:
@markusweimer @afchung So here is what happened:

Earlier all group comm. configuration was added as part of context but since we decided to add in task I moved it to IMRU Driver. Then bugs from evaluator side started coming and all, so I added lots of logs and checks and added all Failure handlerds (IFailedEvaluator, IFailedContext, IFailedTask etc etc. Then on first pull request Julia suggested some unrelated changes and all. So it kept on changing.

dkm2110 commented:
if we feel code is in worse shape than before I can revert. Let me know.

dkm2110 commented:
@afchung I have addressed your comments

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -369,23 +459,33 @@ private void AddGroupCommunicationOperators()
-                _perMapperConfiguration.Push(config);
+                IConfiguration config = _perMapperConfigs.Aggregate(emptyConfig,
+                    (current, configGenerator) =>
+                        Configurations.Merge(current, configGenerator.GetMapperConfiguration(counter, totalMappers)));
```
Is `counter` just `i`?

dkm2110 replied:
fixed

afchung commented:
@dkm2110 left some minor comments. I think this is fine as it is, provided that we've already gone through so many rounds of CR and it would kind of be a waste to revert it. We will need to go through the code again later to clean up the logic though, as @jwang98052 integrates her changes. @markusweimer what do you think?

markusweimer commented:
I'm in favor of merging this, as it unblocks the broader cleanup of IMRU we need to do to support fault tolerance.

jwang98052 commented:
I would like to do one more pass as I haven't got time to look at it again.

dkm2110 commented:
@jwang98052  Please do it soon since this has been there for quite some time with many reviews.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/DataLoadingContext.cs:
```diff
@@ -0,0 +1,65 @@
+
+namespace Org.Apache.REEF.IMRU.OnREEF.Driver
+{
+    internal class DataLoadingContext<T> : IObserver<IContextStart>
```
Document

dkm2110 replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUBugException.cs:
```diff
@@ -0,0 +1,55 @@
+        private static readonly Logger Logger = Logger.GetLogger(typeof(IMRUBugException));
+
+        /// <summary>
+        /// Constructor
```
More descriptive than just a "constructor" even it is a simple class.

dkm2110 replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUBugException.cs:
```diff
@@ -0,0 +1,55 @@
+        }
+
+        /// <summary>
+        /// Constructor
```
More descriptive than just a "constructor"

dkm2110 replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -53,32 +54,41 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
         private readonly bool _invokeGC;
+        private bool _areImruTasksCompleted = false;
+        private int _numberOfReadyTasks = 0;
+        private readonly object _lock = new object();
```
If this lock is only for _completedTasks, you might want to use more specific name like _completedTasksLock

dkm2110 replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,102 +145,46 @@ public void OnNext(IDriverStarted value)
-                _groupCommTaskStarter.QueueTask(partialTaskConf, activeContext);
+                var clientTaskConfig = Configurations.Merge(groupCommConfig, GetUpdateTaskConfiguration());
+                _groupCommTaskStarter.QueueTask(clientTaskConfig, activeContext);
+                RequestMapEvaluators(_totalMappers);
```
Why only request map evaluators after master task is added? activeContext.id to determine which one is for master task, would the sequence of adding tasks matter? 

In fault tolerant scenario, we have to submit all the contexts before adding tasks. Any issue if we make context submitting sequence independent?

dkm2110 replied:
Agreed but this is not in the scope of this JIRA. We can do it as part of FT itself.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUBugException.cs:
```diff
@@ -0,0 +1,55 @@
+    /// Type of exception thrown when possible bugs are detected in IMRU code.
+    /// For example, we reach forbidden region of codes, inconsistent state etc.
+    /// </summary>
+    public sealed class IMRUBugException : Exception
```
Is this IMRU application exception? Rename it as IMRUApplicationException is better otherwise it is like a temp solution.

dkm2110 replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,71 +195,208 @@ public void OnNext(IActiveContext activeContext)
+        /// <param name="value"></param>
+        public void OnNext(IFailedContext value)
+        {
+            if (_areImruTasksCompleted)
```
Do you have test that reaches here?

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,71 +195,208 @@ public void OnNext(IActiveContext activeContext)
-            throw new NotImplementedException("Cannot currently handle exception in OneError", error);
+            if (_areImruTasksCompleted)
+            {
+                Logger.Log(Level.Info,
```
Do you have test that reaches here?

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,71 +195,208 @@ public void OnNext(IActiveContext activeContext)
+        {
+            if (_areImruTasksCompleted)
+            {
+                Logger.Log(Level.Info,
```
Do you have test that reaches this line?

dkm2110 replied:
As discussed these bits of IMRU are going to be replaced very soon. No point writing tests

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,71 +195,208 @@ public void OnNext(IActiveContext activeContext)
+                    string.Format("Context with Id: {0} failed but IMRU task is completed. So ignoring.", value.Id));
+                return;
+            }
+            Exceptions.Throw(new Exception(string.Format("Data Loading Context with Id: {0} failed", value.Id)), Logger);
```
Do you have test that reaches to this code?

dkm2110 replied:
same comment as above

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,71 +195,208 @@ public void OnNext(IActiveContext activeContext)
 
         /// <summary>
-        /// Specfies how to handle exception or error
+        /// Specified what to do if Failed Context is received.
```
Why specified not Specifies?

dkm2110 replied:
fixed

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,71 +195,208 @@ public void OnNext(IActiveContext activeContext)
         /// <param name="completedTask">The link to the completed task</param>
```
typo in the comments Specifies not Specfies

dkm2110 replied:
fixed

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,71 +195,208 @@ public void OnNext(IActiveContext activeContext)
         /// <summary>
-        /// Specfies what to do when driver is done
-        /// In this case do nothing
+        /// Specified what to do once we receive a running task
```
It is inconsistent some where use Specifies some use Specified. I would recommend to use Specifies.

dkm2110 replied:
fixed

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,71 +195,208 @@ public void OnNext(IActiveContext activeContext)
+                    .Build(),
+                    _configurationManager.MapFunctionConfiguration,
+                    mapSpecificConfig)
+                .BindNamedParameter(typeof(InvokeGC), _invokeGC.ToString())
```
use 
.BindNamedParameter<InvokeGC, bool>(GenericType<InvokeGC>.Class, _invokeGC.ToString())

dkm2110 replied:
This is beyond the scope of this jira
On Apr 19, 2016 7:45 PM, "Julia Wang" <notifications@github.com> wrote:

> In lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs
> <https://github.com/apache/reef/pull/952#discussion_r60343817>:
>
> > +            if (!_perMapperConfiguration.TryPop(out mapSpecificConfig))
> > +            {
> > +                Exceptions.Throw(
> > +                    new IMRUBugException(string.Format("No per map configuration exist for the active context {0}",
> > +                        activeContext.Id)),
> > +                    Logger);
> > +            }
> > +
> > +            return TangFactory.GetTang()
> > +                .NewConfigurationBuilder(TaskConfiguration.ConfigurationModule
> > +                    .Set(TaskConfiguration.Identifier, taskId)
> > +                    .Set(TaskConfiguration.Task, GenericType<MapTaskHost<TMapInput, TMapOutput>>.Class)
> > +                    .Build(),
> > +                    _configurationManager.MapFunctionConfiguration,
> > +                    mapSpecificConfig)
> > +                .BindNamedParameter(typeof(InvokeGC), _invokeGC.ToString())
>
> use
> .BindNamedParameter(GenericType.Class, _invokeGC.ToString())
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/reef/pull/952/files/8c300c9c291b4f8cbeda476ff2e20470bd4270e9#r60343817>
>

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -242,71 +195,208 @@ public void OnNext(IActiveContext activeContext)
+                        .Build(),
+                        _configurationManager.UpdateFunctionConfiguration,
+                        _configurationManager.ResultHandlerConfiguration)
+                    .BindNamedParameter(typeof(InvokeGC), _invokeGC.ToString())
```
Use .BindNamedParameter<InvokeGC, bool>(GenericType<InvokeGC>.Class, _invokeGC.ToString())

dkm2110 replied:
done

jwang98052 commented:
I completed a pass.

dkm2110 commented:
@jwang98052 @afchung @markusweimer  I have addressed all the comments.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/DataLoadingContext.cs:
```diff
@@ -0,0 +1,71 @@
+        /// Specifies what to do when context starts.
+        /// </summary>
+        /// <param name="value">context start token</param>
+        /// TODO[REEF-1339] - AddCache() function of IInputPartition will be called here.
```
typo: `AddCache` -> `Add Cache()`

dkm2110 replied:
It seems fine to me...AddCache() is the function name

markusweimer replied:
OK. I suspect it will actually be called `Cache` or something like that. No need to fix it here, though.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUApplicationException.cs:
```diff
@@ -0,0 +1,57 @@
+    /// Type of exception thrown when possible bugs are detected in IMRU code.
+    /// For example, we reach forbidden region of codes, inconsistent state etc.
+    /// </summary>
+    public sealed class IMRUApplicationException : Exception
```
The name is misleading. When I saw the type without the documentation, I assumed this would refer to a bug in an IMRU application, not IMRU itself. Maybe call this `IMRUSystemException`?

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -53,32 +54,41 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
 
         private readonly ConfigurationManager _configurationManager;
-        private readonly IPartitionedInputDataSet _dataSet;
+        private readonly int _totalMappers;
```
Why change these two lines? Don't we still need an `IPartitionedInputDataSet`, which can provide us with the number of mappers we need?

dkm2110 replied:
the whole logic and use of IPartitionedInputDataSet  has gone to ContextAndServiceConfigurationProvider. The only use in driver is to get totalmappers. Hence in driver constructor I use IPartitionedInputDataSet  to assign variable _totalMappers and then never use it in driver.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -53,32 +54,41 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
         private readonly int _memoryForUpdateTask;
         private readonly ISet<IPerMapperConfigGenerator> _perMapperConfigs;
-        private readonly ConcurrentBag<ICompletedTask> _completedTasks;
+        private readonly IList<ICompletedTask> _completedTasks = new List<ICompletedTask>();
```
Isn't `Set` a more logical type for this, as order shouldn't matter?

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -53,32 +54,41 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
         private readonly bool _invokeGC;
+        private bool _areImruTasksCompleted = false;
+        private int _numberOfReadyTasks = 0;
+        private readonly object _completedTasksLock = new object();
```
Can't we use `_completedTasks` as the lock for itself?

dkm2110 replied:
we can and should. done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -53,32 +54,41 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
         private int _currentFailedEvaluators = 0;
-        private bool _reachedUpdateTaskActiveContext = false;
         private readonly bool _invokeGC;
+        private bool _areImruTasksCompleted = false;
```
Can't this be replaced with a computed property that checks `_completedTasks`?

dkm2110 replied:
can be...done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,177 +145,262 @@ public void OnNext(IDriverStarted value)
+                var groupCommConfig = GetGroupCommConfiguration();
                 _commGroup.AddTask(IMRUConstants.UpdateTaskName);
-                _groupCommTaskStarter.QueueTask(partialTaskConf, activeContext);
+                var clientTaskConfig = Configurations.Merge(groupCommConfig, GetUpdateTaskConfiguration());
```
couldn't `GetUpdateTaskConfiguration()` return this merged configuration already?

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,177 +145,262 @@ public void OnNext(IDriverStarted value)
+                    _serviceAndContextConfigurationProvider.GetPartitionIdByEvaluatorId(activeContext.EvaluatorId));
                 _commGroup.AddTask(taskId);
-                _groupCommTaskStarter.QueueTask(partialTaskConf, activeContext);
+                var clientTaskConfig = Configurations.Merge(groupCommConfig,
```
Same as above: Can't `GetMapTaskConfiguration()` include the configuration provided by `GetGroupCommConfiguration()`?

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,177 +145,262 @@ public void OnNext(IDriverStarted value)
+                Logger.Log(Level.Verbose, "Submitting map task");
+                _serviceAndContextConfigurationProvider.RecordActiveContextPerEvaluatorId(activeContext.EvaluatorId);
+                var groupCommConfig = GetGroupCommConfiguration();
+                string taskId = string.Format("{0}-{1}-Version0",
```
Move this to a method of its own.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,177 +145,262 @@ public void OnNext(IDriverStarted value)
+                    string.Format("Received completed task message from task Id: {0}", completedTask.Id));
+                _completedTasks.Add(completedTask);
+
+                if (_completedTasks.Count >= _totalMappers + 1)
```
Move this to a method of its own, e.g. `ShutDownAllEvaluators()` or such.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,177 +145,262 @@ public void OnNext(IDriverStarted value)
+            int currFailedEvaluators = Interlocked.Increment(ref _currentFailedEvaluators);
             if (currFailedEvaluators > _allowedFailedEvaluators)
             {
                 Exceptions.Throw(new Exception("Cannot continue IMRU job, Failed evaluators reach maximum limit"),
```
Add an exception class for this, e.g. `MaximumNumberOfEvalutorFailuresExceededException`. That way, we can report this more orderly to the client.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,177 +145,262 @@ public void OnNext(IDriverStarted value)
+            // evaluator else ask for mapper evaluator
+            if (!isMaster)
             {
+                Logger.Log(Level.Info, "Requesting for the failed map evaluator again");
```
Reword: `"Requesting a replacement Evaluator for {0}", partitionId`

dkm2110 replied:
partitionId or evaluatorId?

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,177 +145,262 @@ public void OnNext(IDriverStarted value)
+        /// Specifies what to do once we receive a running task
         /// </summary>
+        /// <param name="value"></param>
+        public void OnNext(IRunningTask value)
```
This is the same behavior as if IMRU wouldn't bind a handler for `IRunningTask` at all. Probably safe to remove altogether.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -136,177 +145,262 @@ public void OnNext(IDriverStarted value)
+                    .BindImplementation(GenericType<IIMRUResultHandler<TResult>>.Class,
+                        GenericType<DefaultResultHandler<TResult>>.Class)
+                    .Build();
+                Logger.Log(Level.Warning,
```
I'm inclined to make this an `Info`, as the behavior is within the IMRU spec.

dkm2110 replied:
done

markusweimer commented:
I've done another pass, mostly focusing on potential simplifications of the code.

dkm2110 commented:
@markusweimer I have addressed all of your comments. Please have a look.

dkm2110 commented:
@markusweimer @afchung @jwang98052 I have done some pending changes....Let me know if more are required.

markusweimer commented:
This LGTM now. I assume you tested it on YARN?

dkm2110 commented:
Yupp 1000 nodes as usual :)

On Thu, Apr 21, 2016 at 2:43 PM, Markus Weimer <notifications@github.com>
wrote:

> This LGTM now. I assume you tested it on YARN?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/reef/pull/952#issuecomment-213124352>
>

dkm2110 commented:
@markusweimer  Also let me rebase and push and also run all tests once more. I will notify once all this is done and you can merge.

markusweimer commented:
Sounds like a plan! @afchung @jwang98052 Are you OK with a merge?

afchung commented:
+1 on merging.

dkm2110 commented:
@markusweimer  I have done rebase and done testing. As discussed on email thread there are 2 unrelated tests failing only for me and for some too long file reasons. So should be ok. Please merge

jwang98052 commented:
+1 for the merge

markusweimer commented:
Thanks! I will test and merge.
